// @laravel/echo-react/dist/index.js
import K from "pusher-js";
import { useState as N, useEffect as I, useMemo as w, useCallback as u, useRef as S } from "react";
class T {
  constructor() {
    this.notificationCreatedEvent = ".Illuminate\\Notifications\\Events\\BroadcastNotificationCreated";
  }
  /**
   * Listen for a whisper event on the channel instance.
   */
  listenForWhisper(t, e) {
    return this.listen(".client-" + t, e);
  }
  /**
   * Listen for an event on the channel instance.
   */
  notification(t) {
    return this.listen(this.notificationCreatedEvent, t);
  }
  /**
   * Stop listening for notification events on the channel instance.
   */
  stopListeningForNotification(t) {
    return this.stopListening(this.notificationCreatedEvent, t);
  }
  /**
   * Stop listening for a whisper event on the channel instance.
   */
  stopListeningForWhisper(t, e) {
    return this.stopListening(".client-" + t, e);
  }
}
class L {
  /**
   * Create a new class instance.
   */
  constructor(t) {
    this.namespace = t;
  }
  /**
   * Format the given event name.
   */
  format(t) {
    return [".", "\\"].includes(t.charAt(0)) ? t.substring(1) : (this.namespace && (t = this.namespace + "." + t), t.replace(/\./g, "\\"));
  }
  /**
   * Set the event namespace.
   */
  setNamespace(t) {
    this.namespace = t;
  }
}
function U(n) {
  try {
    return Reflect.construct(String, [], n), !0;
  } catch {
    return !1;
  }
}
class P extends T {
  /**
   * Create a new class instance.
   */
  constructor(t, e, s) {
    super(), this.name = e, this.pusher = t, this.options = s, this.eventFormatter = new L(this.options.namespace), this.subscribe();
  }
  /**
   * Subscribe to a Pusher channel.
   */
  subscribe() {
    this.subscription = this.pusher.subscribe(this.name);
  }
  /**
   * Unsubscribe from a Pusher channel.
   */
  unsubscribe() {
    this.pusher.unsubscribe(this.name);
  }
  /**
   * Listen for an event on the channel instance.
   */
  listen(t, e) {
    return this.on(this.eventFormatter.format(t), e), this;
  }
  /**
   * Listen for all events on the channel instance.
   */
  listenToAll(t) {
    return this.subscription.bind_global((e, s) => {
      if (e.startsWith("pusher:"))
        return;
      let r = String(this.options.namespace ?? "").replace(
        /\./g,
        "\\"
      ), i = e.startsWith(r) ? e.substring(r.length + 1) : "." + e;
      t(i, s);
    }), this;
  }
  /**
   * Stop listening for an event on the channel instance.
   */
  stopListening(t, e) {
    return e ? this.subscription.unbind(
      this.eventFormatter.format(t),
      e
    ) : this.subscription.unbind(this.eventFormatter.format(t)), this;
  }
  /**
   * Stop listening for all events on the channel instance.
   */
  stopListeningToAll(t) {
    return t ? this.subscription.unbind_global(t) : this.subscription.unbind_global(), this;
  }
  /**
   * Register a callback to be called anytime a subscription succeeds.
   */
  subscribed(t) {
    return this.on("pusher:subscription_succeeded", () => {
      t();
    }), this;
  }
  /**
   * Register a callback to be called anytime a subscription error occurs.
   */
  error(t) {
    return this.on("pusher:subscription_error", (e) => {
      t(e);
    }), this;
  }
  /**
   * Bind a channel to an event.
   */
  on(t, e) {
    return this.subscription.bind(t, e), this;
  }
}
class j extends P {
  /**
   * Send a whisper event to other clients in the channel.
   */
  whisper(t, e) {
    return this.pusher.channels.channels[this.name].trigger(
      `client-${t}`,
      e
    ), this;
  }
}
class X extends P {
  /**
   * Send a whisper event to other clients in the channel.
   */
  whisper(t, e) {
    return this.pusher.channels.channels[this.name].trigger(
      `client-${t}`,
      e
    ), this;
  }
}
class Q extends j {
  /**
   * Register a callback to be called anytime the member list changes.
   */
  here(t) {
    return this.on("pusher:subscription_succeeded", (e) => {
      t(Object.keys(e.members).map((s) => e.members[s]));
    }), this;
  }
  /**
   * Listen for someone joining the channel.
   */
  joining(t) {
    return this.on("pusher:member_added", (e) => {
      t(e.info);
    }), this;
  }
  /**
   * Send a whisper event to other clients in the channel.
   */
  whisper(t, e) {
    return this.pusher.channels.channels[this.name].trigger(
      `client-${t}`,
      e
    ), this;
  }
  /**
   * Listen for someone leaving the channel.
   */
  leaving(t) {
    return this.on("pusher:member_removed", (e) => {
      t(e.info);
    }), this;
  }
}
class F extends T {
  /**
   * Create a new class instance.
   */
  constructor(t, e, s) {
    super(), this.events = {}, this.listeners = {}, this.name = e, this.socket = t, this.options = s, this.eventFormatter = new L(this.options.namespace), this.subscribe();
  }
  /**
   * Subscribe to a Socket.io channel.
   */
  subscribe() {
    this.socket.emit("subscribe", {
      channel: this.name,
      auth: this.options.auth || {}
    });
  }
  /**
   * Unsubscribe from channel and ubind event callbacks.
   */
  unsubscribe() {
    this.unbind(), this.socket.emit("unsubscribe", {
      channel: this.name,
      auth: this.options.auth || {}
    });
  }
  /**
   * Listen for an event on the channel instance.
   */
  listen(t, e) {
    return this.on(this.eventFormatter.format(t), e), this;
  }
  /**
   * Stop listening for an event on the channel instance.
   */
  stopListening(t, e) {
    return this.unbindEvent(this.eventFormatter.format(t), e), this;
  }
  /**
   * Register a callback to be called anytime a subscription succeeds.
   */
  subscribed(t) {
    return this.on("connect", (e) => {
      t(e);
    }), this;
  }
  /**
   * Register a callback to be called anytime an error occurs.
   */
  error(t) {
    return this;
  }
  /**
   * Bind the channel's socket to an event and store the callback.
   */
  on(t, e) {
    return this.listeners[t] = this.listeners[t] || [], this.events[t] || (this.events[t] = (s, r) => {
      this.name === s && this.listeners[t] && this.listeners[t].forEach((i) => i(r));
    }, this.socket.on(t, this.events[t])), this.listeners[t].push(e), this;
  }
  /**
   * Unbind the channel's socket from all stored event callbacks.
   */
  unbind() {
    Object.keys(this.events).forEach((t) => {
      this.unbindEvent(t);
    });
  }
  /**
   * Unbind the listeners for the given event.
   */
  unbindEvent(t, e) {
    this.listeners[t] = this.listeners[t] || [], e && (this.listeners[t] = this.listeners[t].filter(
      (s) => s !== e
    )), (!e || this.listeners[t].length === 0) && (this.events[t] && (this.socket.removeListener(t, this.events[t]), delete this.events[t]), delete this.listeners[t]);
  }
}
class $ extends F {
  /**
   * Send a whisper event to other clients in the channel.
   */
  whisper(t, e) {
    return this.socket.emit("client event", {
      channel: this.name,
      event: `client-${t}`,
      data: e
    }), this;
  }
}
class W extends $ {
  /**
   * Register a callback to be called anytime the member list changes.
   */
  here(t) {
    return this.on("presence:subscribed", (e) => {
      t(e.map((s) => s.user_info));
    }), this;
  }
  /**
   * Listen for someone joining the channel.
   */
  joining(t) {
    return this.on(
      "presence:joining",
      (e) => t(e.user_info)
    ), this;
  }
  /**
   * Send a whisper event to other clients in the channel.
   */
  whisper(t, e) {
    return this.socket.emit("client event", {
      channel: this.name,
      event: `client-${t}`,
      data: e
    }), this;
  }
  /**
   * Listen for someone leaving the channel.
   */
  leaving(t) {
    return this.on(
      "presence:leaving",
      (e) => t(e.user_info)
    ), this;
  }
}
class C extends T {
  /**
   * Subscribe to a channel.
   */
  subscribe() {
  }
  /**
   * Unsubscribe from a channel.
   */
  unsubscribe() {
  }
  /**
   * Listen for an event on the channel instance.
   */
  listen(t, e) {
    return this;
  }
  /**
   * Listen for all events on the channel instance.
   */
  listenToAll(t) {
    return this;
  }
  /**
   * Stop listening for an event on the channel instance.
   */
  stopListening(t, e) {
    return this;
  }
  /**
   * Register a callback to be called anytime a subscription succeeds.
   */
  subscribed(t) {
    return this;
  }
  /**
   * Register a callback to be called anytime an error occurs.
   */
  error(t) {
    return this;
  }
  /**
   * Bind a channel to an event.
   */
  on(t, e) {
    return this;
  }
}
class H extends C {
  /**
   * Send a whisper event to other clients in the channel.
   */
  whisper(t, e) {
    return this;
  }
}
class z extends C {
  /**
   * Send a whisper event to other clients in the channel.
   */
  whisper(t, e) {
    return this;
  }
}
class Y extends H {
  /**
   * Register a callback to be called anytime the member list changes.
   */
  here(t) {
    return this;
  }
  /**
   * Listen for someone joining the channel.
   */
  joining(t) {
    return this;
  }
  /**
   * Send a whisper event to other clients in the channel.
   */
  whisper(t, e) {
    return this;
  }
  /**
   * Listen for someone leaving the channel.
   */
  leaving(t) {
    return this;
  }
}
const B = class q {
  /**
   * Create a new class instance.
   */
  constructor(t) {
    this.setOptions(t), this.connect();
  }
  /**
   * Merge the custom options with the defaults.
   */
  setOptions(t) {
    this.options = {
      ...q._defaultOptions,
      ...t,
      broadcaster: t.broadcaster
    };
    let e = this.csrfToken();
    e && (this.options.auth.headers["X-CSRF-TOKEN"] = e, this.options.userAuthentication.headers["X-CSRF-TOKEN"] = e), e = this.options.bearerToken, e && (this.options.auth.headers.Authorization = "Bearer " + e, this.options.userAuthentication.headers.Authorization = "Bearer " + e);
  }
  /**
   * Extract the CSRF token from the page.
   */
  csrfToken() {
    var t, e;
    return typeof window < "u" && (t = window.Laravel) != null && t.csrfToken ? window.Laravel.csrfToken : this.options.csrfToken ? this.options.csrfToken : typeof document < "u" && typeof document.querySelector == "function" ? ((e = document.querySelector('meta[name="csrf-token"]')) == null ? void 0 : e.getAttribute("content")) ?? null : null;
  }
};
B._defaultOptions = {
  auth: {
    headers: {}
  },
  authEndpoint: "/broadcasting/auth",
  userAuthentication: {
    endpoint: "/broadcasting/user-auth",
    headers: {}
  },
  csrfToken: null,
  bearerToken: null,
  host: null,
  key: null,
  namespace: "App.Events"
};
let R = B;
class y extends R {
  constructor() {
    super(...arguments), this.channels = {};
  }
  /**
   * Create a fresh Pusher connection.
   */
  connect() {
    if (typeof this.options.client < "u")
      this.pusher = this.options.client;
    else if (this.options.Pusher)
      this.pusher = new this.options.Pusher(
        this.options.key,
        this.options
      );
    else if (typeof window < "u" && typeof window.Pusher < "u")
      this.pusher = new window.Pusher(this.options.key, this.options);
    else
      throw new Error(
        "Pusher client not found. Should be globally available or passed via options.client"
      );
  }
  /**
   * Sign in the user via Pusher user authentication (https://pusher.com/docs/channels/using_channels/user-authentication/).
   */
  signin() {
    this.pusher.signin();
  }
  /**
   * Listen for an event on a channel instance.
   */
  listen(t, e, s) {
    return this.channel(t).listen(e, s);
  }
  /**
   * Get a channel instance by name.
   */
  channel(t) {
    return this.channels[t] || (this.channels[t] = new P(
      this.pusher,
      t,
      this.options
    )), this.channels[t];
  }
  /**
   * Get a private channel instance by name.
   */
  privateChannel(t) {
    return this.channels["private-" + t] || (this.channels["private-" + t] = new j(
      this.pusher,
      "private-" + t,
      this.options
    )), this.channels["private-" + t];
  }
  /**
   * Get a private encrypted channel instance by name.
   */
  encryptedPrivateChannel(t) {
    return this.channels["private-encrypted-" + t] || (this.channels["private-encrypted-" + t] = new X(
      this.pusher,
      "private-encrypted-" + t,
      this.options
    )), this.channels["private-encrypted-" + t];
  }
  /**
   * Get a presence channel instance by name.
   */
  presenceChannel(t) {
    return this.channels["presence-" + t] || (this.channels["presence-" + t] = new Q(
      this.pusher,
      "presence-" + t,
      this.options
    )), this.channels["presence-" + t];
  }
  /**
   * Leave the given channel, as well as its private and presence variants.
   */
  leave(t) {
    [
      t,
      "private-" + t,
      "private-encrypted-" + t,
      "presence-" + t
    ].forEach((e) => {
      this.leaveChannel(e);
    });
  }
  /**
   * Leave the given channel.
   */
  leaveChannel(t) {
    this.channels[t] && (this.channels[t].unsubscribe(), delete this.channels[t]);
  }
  /**
   * Get the socket ID for the connection.
   */
  socketId() {
    return this.pusher.connection.socket_id;
  }
  /**
   * Get the current connection status.
   */
  connectionStatus() {
    const t = this.pusher.connection.state;
    switch (t) {
      case "connected":
      case "connecting":
        return t;
      case "failed":
      case "unavailable":
        return "failed";
      default:
        return "disconnected";
    }
  }
  /**
   * Subscribe to connection status changes.
   */
  onConnectionChange(t) {
    const e = () => {
      t(this.connectionStatus());
    }, s = ["state_change", "connected", "disconnected"];
    return s.forEach((r) => {
      this.pusher.connection.bind(r, e);
    }), () => {
      s.forEach((r) => {
        this.pusher.connection.unbind(r, e);
      });
    };
  }
  /**
   * Disconnect Pusher connection.
   */
  disconnect() {
    this.pusher.disconnect();
  }
}
class M extends R {
  constructor() {
    super(...arguments), this.channels = {};
  }
  /**
   * Create a fresh Socket.io connection.
   */
  connect() {
    let t = this.getSocketIO();
    this.socket = t(
      this.options.host ?? void 0,
      this.options
    ), this.socket.io.on("reconnect", () => {
      Object.values(this.channels).forEach((e) => {
        e.subscribe();
      });
    });
  }
  /**
   * Get socket.io module from global scope or options.
   */
  getSocketIO() {
    if (typeof this.options.client < "u")
      return this.options.client;
    if (typeof window < "u" && typeof window.io < "u")
      return window.io;
    throw new Error(
      "Socket.io client not found. Should be globally available or passed via options.client"
    );
  }
  /**
   * Listen for an event on a channel instance.
   */
  listen(t, e, s) {
    return this.channel(t).listen(e, s);
  }
  /**
   * Get a channel instance by name.
   */
  channel(t) {
    return this.channels[t] || (this.channels[t] = new F(
      this.socket,
      t,
      this.options
    )), this.channels[t];
  }
  /**
   * Get a private channel instance by name.
   */
  privateChannel(t) {
    return this.channels["private-" + t] || (this.channels["private-" + t] = new $(
      this.socket,
      "private-" + t,
      this.options
    )), this.channels["private-" + t];
  }
  /**
   * Get a presence channel instance by name.
   */
  presenceChannel(t) {
    return this.channels["presence-" + t] || (this.channels["presence-" + t] = new W(
      this.socket,
      "presence-" + t,
      this.options
    )), this.channels["presence-" + t];
  }
  /**
   * Leave the given channel, as well as its private and presence variants.
   */
  leave(t) {
    [t, "private-" + t, "presence-" + t].forEach((e) => {
      this.leaveChannel(e);
    });
  }
  /**
   * Leave the given channel.
   */
  leaveChannel(t) {
    this.channels[t] && (this.channels[t].unsubscribe(), delete this.channels[t]);
  }
  /**
   * Get the socket ID for the connection.
   */
  socketId() {
    return this.socket.id;
  }
  /**
   * Get the current connection status.
   */
  connectionStatus() {
    return this.socket.connected ? "connected" : this.socket.io._reconnecting ? "reconnecting" : this.socket.id !== void 0 ? "disconnected" : "connecting";
  }
  /**
   * Subscribe to connection status changes.
   */
  onConnectionChange(t) {
    const e = () => {
      t(this.connectionStatus());
    }, s = [
      "connect",
      "disconnect",
      "connect_error",
      "reconnect_attempt",
      "reconnect",
      "reconnect_error",
      "reconnect_failed"
    ];
    return s.forEach((r) => {
      this.socket.on(r, e);
    }), () => {
      s.forEach((r) => {
        this.socket.off(r, e);
      });
    };
  }
  /**
   * Disconnect Socketio connection.
   */
  disconnect() {
    this.socket.disconnect();
  }
}
class O extends R {
  constructor() {
    super(...arguments), this.channels = {};
  }
  /**
   * Create a fresh connection.
   */
  connect() {
  }
  /**
   * Listen for an event on a channel instance.
   */
  listen(t, e, s) {
    return new C();
  }
  /**
   * Get a channel instance by name.
   */
  channel(t) {
    return new C();
  }
  /**
   * Get a private channel instance by name.
   */
  privateChannel(t) {
    return new H();
  }
  /**
   * Get a private encrypted channel instance by name.
   */
  encryptedPrivateChannel(t) {
    return new z();
  }
  /**
   * Get a presence channel instance by name.
   */
  presenceChannel(t) {
    return new Y();
  }
  /**
   * Leave the given channel, as well as its private and presence variants.
   */
  leave(t) {
  }
  /**
   * Leave the given channel.
   */
  leaveChannel(t) {
  }
  /**
   * Get the socket ID for the connection.
   */
  socketId() {
    return "fake-socket-id";
  }
  /**
   * Get the current connection status.
   */
  connectionStatus() {
    return "connected";
  }
  /**
   * Subscribe to connection status changes.
   */
  onConnectionChange(t) {
    return () => {
    };
  }
  /**
   * Disconnect the connection.
   */
  disconnect() {
  }
}
class D {
  /**
   * Create a new class instance.
   */
  constructor(t) {
    this.options = t, this.connect(), this.options.withoutInterceptors || this.registerInterceptors();
  }
  /**
   * Get a channel instance by name.
   */
  channel(t) {
    return this.connector.channel(t);
  }
  /**
   * Create a new connection.
   */
  connect() {
    if (this.options.broadcaster === "reverb")
      this.connector = new y({
        ...this.options,
        cluster: ""
      });
    else if (this.options.broadcaster === "pusher")
      this.connector = new y(this.options);
    else if (this.options.broadcaster === "ably")
      this.connector = new y({
        ...this.options,
        cluster: "",
        broadcaster: "pusher"
      });
    else if (this.options.broadcaster === "socket.io")
      this.connector = new M(this.options);
    else if (this.options.broadcaster === "null")
      this.connector = new O(this.options);
    else if (typeof this.options.broadcaster == "function" && U(this.options.broadcaster))
      this.connector = new this.options.broadcaster(this.options);
    else
      throw new Error(
        `Broadcaster ${typeof this.options.broadcaster} ${String(this.options.broadcaster)} is not supported.`
      );
  }
  /**
   * Disconnect from the Echo server.
   */
  disconnect() {
    this.connector.disconnect();
  }
  /**
   * Get a presence channel instance by name.
   */
  join(t) {
    return this.connector.presenceChannel(t);
  }
  /**
   * Leave the given channel, as well as its private and presence variants.
   */
  leave(t) {
    this.connector.leave(t);
  }
  /**
   * Leave the given channel.
   */
  leaveChannel(t) {
    this.connector.leaveChannel(t);
  }
  /**
   * Leave all channels.
   */
  leaveAllChannels() {
    for (const t in this.connector.channels)
      this.leaveChannel(t);
  }
  /**
   * Listen for an event on a channel instance.
   */
  listen(t, e, s) {
    return this.connector.listen(t, e, s);
  }
  /**
   * Get a private channel instance by name.
   */
  private(t) {
    return this.connector.privateChannel(t);
  }
  /**
   * Get a private encrypted channel instance by name.
   */
  encryptedPrivate(t) {
    if (this.connectorSupportsEncryptedPrivateChannels(this.connector))
      return this.connector.encryptedPrivateChannel(t);
    throw new Error(
      `Broadcaster ${typeof this.options.broadcaster} ${String(
        this.options.broadcaster
      )} does not support encrypted private channels.`
    );
  }
  connectorSupportsEncryptedPrivateChannels(t) {
    return t instanceof y || t instanceof O;
  }
  /**
   * Get the Socket ID for the connection.
   */
  socketId() {
    return this.connector.socketId();
  }
  /**
   * Get the current connection status.
   */
  connectionStatus() {
    return this.connector.connectionStatus();
  }
  /**
   * Register 3rd party request interceptors. These are used to automatically
   * send a connections socket id to a Laravel app with a X-Socket-Id header.
   */
  registerInterceptors() {
    typeof Vue < "u" && Vue != null && Vue.http && this.registerVueRequestInterceptor(), typeof axios == "function" && this.registerAxiosRequestInterceptor(), typeof jQuery == "function" && this.registerjQueryAjaxSetup(), typeof Turbo == "object" && this.registerTurboRequestInterceptor();
  }
  /**
   * Register a Vue HTTP interceptor to add the X-Socket-ID header.
   */
  registerVueRequestInterceptor() {
    Vue.http.interceptors.push(
      (t, e) => {
        this.socketId() && t.headers.set("X-Socket-ID", this.socketId()), e();
      }
    );
  }
  /**
   * Register an Axios HTTP interceptor to add the X-Socket-ID header.
   */
  registerAxiosRequestInterceptor() {
    axios.interceptors.request.use(
      (t) => (this.socketId() && (t.headers["X-Socket-Id"] = this.socketId()), t)
    );
  }
  /**
   * Register jQuery AjaxPrefilter to add the X-Socket-ID header.
   */
  registerjQueryAjaxSetup() {
    typeof jQuery.ajax < "u" && jQuery.ajaxPrefilter(
      (t, e, s) => {
        this.socketId() && s.setRequestHeader("X-Socket-Id", this.socketId());
      }
    );
  }
  /**
   * Register the Turbo Request interceptor to add the X-Socket-ID header.
   */
  registerTurboRequestInterceptor() {
    document.addEventListener(
      "turbo:before-fetch-request",
      (t) => {
        t.detail.fetchOptions.headers["X-Socket-Id"] = this.socketId();
      }
    );
  }
}
let v = null, d = null;
const J = () => {
  if (v)
    return v;
  if (!d)
    throw new Error(
      "Echo has not been configured. Please call `configureEcho()`."
    );
  return d.Pusher ?? (d.Pusher = K), v = new D(d), v;
}, nt = (n) => {
  d = {
    ...{
      reverb: {
        broadcaster: "reverb",
        key: (typeof import.meta.env !== 'undefined' ? import.meta.env.VITE_REVERB_APP_KEY : undefined),
        wsHost: (typeof import.meta.env !== 'undefined' ? import.meta.env.VITE_REVERB_HOST : undefined),
        wsPort: (typeof import.meta.env !== 'undefined' ? import.meta.env.VITE_REVERB_PORT : undefined),
        wssPort: (typeof import.meta.env !== 'undefined' ? import.meta.env.VITE_REVERB_PORT : undefined),
        forceTLS: ((typeof import.meta.env !== 'undefined' ? import.meta.env.VITE_REVERB_SCHEME : undefined) ?? "https") === "https",
        enabledTransports: ["ws", "wss"]
      },
      pusher: {
        broadcaster: "pusher",
        key: (typeof import.meta.env !== 'undefined' ? import.meta.env.VITE_PUSHER_APP_KEY : undefined),
        cluster: (typeof import.meta.env !== 'undefined' ? import.meta.env.VITE_PUSHER_APP_CLUSTER : undefined),
        forceTLS: !0,
        wsHost: (typeof import.meta.env !== 'undefined' ? import.meta.env.VITE_PUSHER_HOST : undefined),
        wsPort: (typeof import.meta.env !== 'undefined' ? import.meta.env.VITE_PUSHER_PORT : undefined),
        wssPort: (typeof import.meta.env !== 'undefined' ? import.meta.env.VITE_PUSHER_PORT : undefined),
        enabledTransports: ["ws", "wss"]
      },
      "socket.io": {
        broadcaster: "socket.io",
        host: (typeof import.meta.env !== 'undefined' ? import.meta.env.VITE_SOCKET_IO_HOST : undefined)
      },
      null: {
        broadcaster: "null"
      },
      ably: {
        broadcaster: "pusher",
        key: (typeof import.meta.env !== 'undefined' ? import.meta.env.VITE_ABLY_PUBLIC_KEY : undefined),
        wsHost: "realtime-pusher.ably.io",
        wsPort: 443,
        disableStats: !0,
        encrypted: !0
      }
    }[n.broadcaster],
    ...n
  }, v && (v = null);
}, E = () => J(), st = () => d !== null, A = (n) => Array.isArray(n) ? n : [n], p = {}, G = (n) => {
  const t = E();
  return n.visibility === "presence" ? t.join(n.name) : n.visibility === "private" ? t.private(n.name) : t.channel(n.name);
}, Z = (n, t) => {
  p[n.id] && (p[n.id].count -= 1, !(p[n.id].count > 0) && (t ? E().leave(n.name) : E().leaveChannel(n.id), delete p[n.id]));
}, V = (n) => {
  if (p[n.id])
    return p[n.id].count += 1, p[n.id].connection;
  const t = G(n);
  return p[n.id] = {
    count: 1,
    connection: t
  }, t;
};
function _(n, t = [], e = () => {
}, s = [], r = "private") {
  const i = w(
    () => ({
      name: n,
      id: ["private", "presence"].includes(r) ? `${r}-${n}` : n,
      visibility: r
    }),
    [n, r]
  ), l = u(e, s), c = S(!1), m = S(!1), h = S(
    V(i)
  ), f = Array.isArray(t) ? JSON.stringify(t) : t, a = w(() => A(t), [f]), o = u(() => {
    c.current && (a.forEach((g) => {
      h.current.stopListening(g, l);
    }), c.current = !1);
  }, [a, l]), k = u(() => {
    c.current || (a.forEach((g) => {
      h.current.listen(g, l);
    }), c.current = !0);
  }, [a, l]), b = u(
    (g = !1) => {
      o(), Z(i, g);
    },
    [o, i]
  ), x = u(() => {
    b(!0);
  }, [b]);
  return I(() => (m.current && (h.current = V(i)), m.current = !0, k(), b), [k, b, i]), w(
    () => ({
      /**
       * Leave the channel
       */
      leaveChannel: b,
      /**
       * Leave the channel and also its associated private and presence channels
       */
      leave: x,
      /**
       * Stop listening for event(s) without leaving the channel
       */
      stopListening: o,
      /**
       * Listen for event(s)
       */
      listen: k,
      /**
       * Channel instance
       */
      channel: () => h.current
    }),
    [x, k, o, b]
  );
}
const rt = (n, t = () => {
}, e = [], s = []) => {
  const r = _(
    n,
    [],
    t,
    s,
    "private"
  ), i = Array.isArray(e) ? JSON.stringify(e) : e, l = w(() => A(e).map((o) => o.includes(".") ? [o, o.replace(/\./g, "\\")] : [o, o.replace(/\\/g, ".")]).flat(), [i]), c = S(!1), m = u(t, s), h = u(
    (o) => {
      c.current && (l.length === 0 || l.includes(o.type)) && m(o);
    },
    [m, l]
  ), f = u(() => {
    c.current || (r.channel().notification(h), c.current = !0);
  }, [h, r]), a = u(() => {
    c.current && (r.channel().stopListeningForNotification(h), c.current = !1);
  }, [h, r]);
  return I(() => (f(), () => a()), [f, a]), w(
    () => ({
      ...r,
      /**
       * Stop listening for notification events
       */
      stopListening: a,
      /**
       * Listen for notification events
       */
      listen: f
    }),
    [r, a, f]
  );
}, it = (n, t = [], e = () => {
}, s = []) => _(
  n,
  t,
  e,
  s,
  "presence"
), ot = (n, t = [], e = () => {
}, s = []) => _(
  n,
  t,
  e,
  s,
  "public"
), ct = (n, t, e = [], s = () => {
}, r = []) => _(
  `${n}.${t}`,
  A(e).map((i) => i.startsWith(".") ? i : `.${i}`),
  s,
  r,
  "private"
), ht = () => {
  const [n, t] = N(
    () => E().connectionStatus()
  );
  return I(() => E().connector.onConnectionChange(
    (e) => {
      t(e);
    }
  ), []), n;
};
export {
  nt as configureEcho,
  E as echo,
  st as echoIsConfigured,
  ht as useConnectionStatus,
  _ as useEcho,
  ct as useEchoModel,
  rt as useEchoNotification,
  it as useEchoPresence,
  ot as useEchoPublic
};
